File: Loenn/entities/gravityDreamBlock.lua
-- Copyright (c) Shane Woolcock. Licensed under the MIT Licence.
-- See the LICENCE file in the repository root for full licence text.

local consts = require("mods").requireFromPlugin("consts")
local helpers = require("mods").requireFromPlugin("helpers")
local utils = require("utils")
--local colors = require("consts.xna_colors")

local drawableSprite = require("structs.drawable_sprite")
local drawableFunction = require("structs.drawable_function")

local doubleArrowTexture = "objects/GravityHelper/gravityDreamBlock/doubleArrow"
local doubleArrowSmallTexture = "objects/GravityHelper/gravityDreamBlock/doubleArrowSmall"
local upArrowTexture = "objects/GravityHelper/gravityDreamBlock/upArrow"
local upArrowSmallTexture = "objects/GravityHelper/gravityDreamBlock/upArrowSmall"
local downArrowTexture = "objects/GravityHelper/gravityDreamBlock/downArrow"
local downArrowSmallTexture = "objects/GravityHelper/gravityDreamBlock/downArrowSmall"

local placementData = helpers.createPlacementData('1', {
    width = 8,
    height = 8,
    fastMoving = false,
    oneUse = false,
    below = false,
    gravityType = consts.gravityTypes.normal.index,
    lineColor = "",
    backColor = "",
    particleColor = "",
    fallType = 0,
    climbFall = true,
    endFallOnSolidTiles = true,
    invertFallingDirFlag = "",
    swapType = 0,
})

local gravityDreamBlock = {
    name = "GravityHelper/GravityDreamBlock",
    ignoredFields = consts.ignoredFields,
    fieldInformation = {
        gravityType = consts.fieldInformation.gravityType(),
        fallType = consts.fieldInformation.fallType,
        swapType = consts.fieldInformation.swapType,
    },
    placements = {
        {
            name = "normal",
            data = helpers.union(placementData, {
                gravityType = consts.gravityTypes.normal.index,
            }),
        },
        {
            name = "inverted",
            data = helpers.union(placementData, {
                gravityType = consts.gravityTypes.inverted.index,
            }),
        },
        {
            name = "toggle",
            data = helpers.union(placementData, {
                gravityType = consts.gravityTypes.toggle.index,
            }),
        },
        {
            name = "normal_swap",
            data = helpers.union(placementData, {
                gravityType = consts.gravityTypes.normal.index,
                swapType = 1,
            }),
        },
        {
            name = "inverted_swap",
            data = helpers.union(placementData, {
                gravityType = consts.gravityTypes.inverted.index,
                swapType = 1,
            }),
        },
        {
            name = "toggle_swap",
            data = helpers.union(placementData, {
                gravityType = consts.gravityTypes.toggle.index,
                swapType = 1,
            }),
        },
    },
}

local function lightened(color, lightness, alpha)
    return {
        math.max(0, math.min(1, color[1] + lightness)),
        math.max(0, math.min(1, color[2] + lightness)),
        math.max(0, math.min(1, color[3] + lightness)),
        alpha or color[4]
    }
end

function gravityDreamBlock.depth(room, entity)
    return entity.below and 4999 or -11001
end

local function putInside(entity, x, y, node)
    local ex = node and node.x or entity.x
    local ey = node and node.y or entity.y

    local right, bottom = ex + entity.width, ey + entity.height
    if x > right then
        x = x - math.ceil((x - right) / entity.width) * entity.width
    elseif x < ex then
        x = x + math.ceil((ex - x) / entity.width) * entity.width
    end
    if y > bottom then
        y = y - math.ceil((y - bottom) / entity.height) * entity.height
    elseif y < ey then
        y = y + math.ceil((ey - y) / entity.height) * entity.height
    end
    return x, y
end

local function mainSprite(room, entity, node)
    -- find sprites
    local largeSprite =
    entity.gravityType == 0 and drawableSprite.fromTexture(downArrowTexture) or
            entity.gravityType == 1 and drawableSprite.fromTexture(upArrowTexture) or
            drawableSprite.fromTexture(doubleArrowTexture)
    local smallSprite =
    entity.gravityType == 0 and drawableSprite.fromTexture(downArrowSmallTexture) or
            entity.gravityType == 1 and drawableSprite.fromTexture(upArrowSmallTexture) or
            drawableSprite.fromTexture(doubleArrowSmallTexture)

    -- custom random so we don't poison the global seed
    local seed = 0
    local function rand()
        local a = 1664525
        local c = 1013904223
        local m = 2^32
        seed = (a * seed + c) % m
        return seed / m
    end

    local spr = drawableFunction.fromFunction(function()
        local gravityColor = consts.gravityTypeForIndex(entity.gravityType).color
        local lineColor, backColor, particleColor = lightened(gravityColor, 0.4), lightened(gravityColor, 0.4), gravityColor
        local parsed, r, g, b, a, oldAlpha
        local ex,ey = node and node.x or entity.x, node and node.y or entity.y

        parsed, r, g, b = utils.parseHexColor(entity.lineColor or "")
        if parsed then lineColor = {r, g, b} end

        parsed, r, g, b = utils.parseHexColor(entity.backColor or "")
        if parsed then backColor = {r, g, b} end

        parsed, r, g, b = utils.parseHexColor(entity.particleColor or "")
        if parsed then particleColor = {r, g, b} end

        -- fill
        r, g, b, oldAlpha = love.graphics.getColor()
        a = oldAlpha * (node and 0.75 or 1)

        love.graphics.setColor(0, 0, 0, a)
        love.graphics.rectangle("fill", ex, ey, entity.width, entity.height)
        love.graphics.setColor(backColor[1], backColor[2], backColor[3], a * 0.12)
        love.graphics.rectangle("fill", ex, ey, entity.width, entity.height)

        -- set scissor for particles
        local sx, sy, sw, sh = love.graphics.getScissor()
        local x1, y1 = love.graphics.transformPoint(ex, ey)
        local x2, y2 = love.graphics.transformPoint(ex + entity.width, ey + entity.height)
        love.graphics.setScissor(x1, y1, x2 - x1, y2 - y1)

        -- set seed
        seed = entity._id or 1

        -- render particles
        local particleCount = math.floor((entity.width / 8) * (entity.height / 8) * 0.7)
        for _ = 1,particleCount do
            -- get random values
            local px,py = math.floor(rand() * entity.width), math.floor(rand() * entity.height)
            local lightness = rand() - 0.25
            local layer = math.floor(rand() * 6)
            -- calculate visuals
            local adjusted = lightened(particleColor, lightness, 0.8)
            adjusted[4] *= a

            px, py = putInside(entity, px, py, node)

            if layer < 3 then
                local particleSprite = layer < 1 and largeSprite or smallSprite
                particleSprite.x = px
                particleSprite.y = py
                particleSprite.color = adjusted
                particleSprite:draw()
            else
                love.graphics.setColor(adjusted)
                love.graphics.rectangle("fill", px, py, 1, 1)
            end
        end

        -- reset scissor
        love.graphics.setScissor(sx, sy, sw, sh)

        -- line
        love.graphics.setColor(lineColor[1], lineColor[2], lineColor[3], a)
        love.graphics.rectangle("line", ex + 0.5, ey + 0.5, entity.width - 1, entity.height - 1)

        -- reset color
        love.graphics.setColor(r, g, b, oldAlpha)
    end)
    return spr
end

function gravityDreamBlock.sprite(room, entity)
    local sprites = {}
    local spr = mainSprite(room, entity, nil)
    table.insert(sprites, spr)
    if entity.swapType and entity.swapType > 0 then
        helpers.addSwapTrailSprites(sprites, entity)
    end
    return sprites
end

function gravityDreamBlock.nodeSprite(room, entity, node, nodeIndex)
    local sprites = {}
    local spr = mainSprite(room, entity, node)
    table.insert(sprites, spr)
    return sprites
end

function gravityDreamBlock.nodeLimits(room, entity)
    if entity.swapType and entity.swapType > 0 then
        return {1, 1}
    else
        return {0, 0}
    end
end

function gravityDreamBlock.selection(room, entity)
    local nodes = entity.nodes
    local x, y = entity.x or 0, entity.y or 0
    local nodeX, nodeY = nodes and nodes[1].x or x, nodes and nodes[1].y or y
    local width, height = entity.width or 8, entity.height or 8

    return utils.rectangle(x, y, width, height), {utils.rectangle(nodeX, nodeY, width, height)}
end

return gravityDreamBlock

File: Ahorn/entities/gravityDreamBlock.jl
# Copyright (c) Shane Woolcock. Licensed under the MIT Licence.
# See the LICENCE file in the repository root for full licence text.

module GravityHelperGravityDreamBlock

using ..Ahorn, Maple

const PLUGIN_VERSION = "1"

@mapdef Entity "GravityHelper/GravityDreamBlock" GravityDreamBlock(
    x::Integer, y::Integer,
    width::Integer=8, height::Integer=8,
    pluginVersion::String=PLUGIN_VERSION,
    fastMoving::Bool=false, oneUse::Bool=false, below::Bool=false,
    gravityType::Integer=0,
    lineColor::String="", backColor::String="", particleColor::String="",
    fall::Bool=false, climbFall::Bool=true
)

const placements = Ahorn.PlacementDict(
    "Gravity Dream Block (Normal) (GravityHelper)" => Ahorn.EntityPlacement(
        GravityDreamBlock,
        "rectangle",
        Dict{String, Any}(
            "gravityType" => 0,
        )
    ),
    "Gravity Dream Block (Inverted) (GravityHelper)" => Ahorn.EntityPlacement(
        GravityDreamBlock,
        "rectangle",
        Dict{String, Any}(
            "gravityType" => 1,
        )
    ),
    "Gravity Dream Block (Toggle) (GravityHelper)" => Ahorn.EntityPlacement(
        GravityDreamBlock,
        "rectangle",
        Dict{String, Any}(
            "gravityType" => 2,
        )
    )
)

const gravityTypes = Dict{String, Integer}(
    "Normal" => 0,
    "Inverted" => 1,
    "Toggle" => 2,
)

const gravityColors = Dict{Integer, Tuple{Real, Real, Real, Real}}(
    0 => (0.0, 0.0, 1.0, 0.5),
    1 => (1.0, 0.0, 0.0, 0.5),
    2 => (0.5, 0.0, 0.5, 0.5),
)

Ahorn.editingIgnored(entity::GravityDreamBlock, multiple::Bool=false) = multiple ? String["x", "y", "modVersion", "pluginVersion"] : String["modVersion", "pluginVersion"]

Ahorn.editingOptions(entity::GravityDreamBlock) = Dict{String, Any}(
    "gravityType" => gravityTypes,
)

Ahorn.nodeLimits(entity::GravityDreamBlock) = 0, 1
Ahorn.minimumSize(entity::GravityDreamBlock) = 8, 8
Ahorn.resizable(entity::GravityDreamBlock) = true, true

upArrowSprite = "objects/GravityHelper/gravityDreamBlock/upArrow"
downArrowSprite = "objects/GravityHelper/gravityDreamBlock/downArrow"
doubleArrowSprite = "objects/GravityHelper/gravityDreamBlock/doubleArrow"

function Ahorn.selection(entity::GravityDreamBlock)
    x, y = Ahorn.position(entity)
    width = Int(get(entity.data, "width", 8))
    height = Int(get(entity.data, "height", 8))
    nodes = get(entity.data, "nodes", ())

    if isempty(nodes)
        return Ahorn.Rectangle(x, y, width, height)
    else
        nx, ny = Int.(nodes[1])
        return [Ahorn.Rectangle(x, y, width, height), Ahorn.Rectangle(nx, ny, width, height)]
    end
end

function lightened(color::Tuple{Real, Real, Real, Real}, amount::Real)
    return (clamp(color[0] + amount, 0, 1), clamp(color[1] + amount, 0, 1), clamp(color[2] + amount, 0, 1), color[3])
end

function renderSpaceJam(ctx::Ahorn.Cairo.CairoContext, entity::GravityDreamBlock, x::Number, y::Number, width::Number, height::Number)
    Ahorn.Cairo.save(ctx)

    Ahorn.set_antialias(ctx, 1)
    Ahorn.set_line_width(ctx, 1)

    gravityType = Int(get(entity.data, "gravityType", 0))
    sprite = gravityType == 0 ? downArrowSprite : gravityType == 1 ? upArrowSprite : doubleArrowSprite
    gravityColor = gravityColors[gravityType]
    entityLineColor = get(entity.data, "lineColor", "")
    entityBackColor = get(entity.data, "backColor", "")
    entityParticleColor = get(entity.data, "particleColor", "")
    lineColor = clamp.(gravityColor .+ 0.4, 0.0, 1.0)
    fillColor = lineColor
    particleColor = lineColor
    
    parsedLineColor = parseColor(entityLineColor)
    parsedBackColor = parseColor(entityBackColor)
    parsedParticleColor = parseColor(entityParticleColor)

    if parsedLineColor != nothing
        lineColor = parsedLineColor
    end

    if parsedBackColor != nothing
        fillColor = parsedBackColor
    end
    
    if parsedParticleColor != nothing
        particleColor = parsedParticleColor
    end
    
    fillColor = (fillColor[1] * 0.12, fillColor[2] * 0.12, fillColor[3] * 0.12, 1.0)
    particleColor = (particleColor[1], particleColor[2], particleColor[3], 1.0)
    
    Ahorn.drawRectangle(ctx, x, y, width, height, fillColor, (lineColor[1], lineColor[2], lineColor[3], 1.0))

    Ahorn.drawSprite(ctx, sprite, x + width / 2, y + height / 2, tint=particleColor)
    Ahorn.restore(ctx)
end

function Ahorn.renderSelectedAbs(ctx::Ahorn.Cairo.CairoContext, entity::GravityDreamBlock)
    x, y = Ahorn.position(entity)
    nodes = get(entity.data, "nodes", ())
    width = Int(get(entity.data, "width", 8))
    height = Int(get(entity.data, "height", 8))

    if !isempty(nodes)
        nx, ny = Int.(nodes[1])
        cox, coy = floor(Int, width / 2), floor(Int, height / 2)
        renderSpaceJam(ctx, entity, nx, ny, width, height)
        Ahorn.drawArrow(ctx, x + cox, y + coy, nx + cox, ny + coy, Ahorn.colors.selection_selected_fc, headLength=6)
    end
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::GravityDreamBlock, room::Maple.Room)
    x = Int(get(entity.data, "x", 0))
    y = Int(get(entity.data, "y", 0))

    width = Int(get(entity.data, "width", 32))
    height = Int(get(entity.data, "height", 32))

    renderSpaceJam(ctx, entity, 0, 0, width, height)
end

function parseColor(hex::String)
    if length(hex) != 6
        return nothing
    end
    parsed = tryparse(Int, hex, base=16)
    return parsed != nothing ? Ahorn.argb32ToRGBATuple(parsed)[1:3] ./ 255 : nothing
end

end